spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/HeaderBasedMessagingInjector.java

```diff
@@ -0,0 +1,123 @@
+package org.springframework.cloud.sleuth.instrument.messaging;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import org.springframework.cloud.sleuth.*;
+import org.springframework.util.StringUtils;
+
+/**
+ * Default implementation for messaging
+ *
+ * @author Marcin Grzejszczak
+ * @since 1.2.0
+ */
+public class HeaderBasedMessagingInjector implements MessagingSpanTextMapInjector {
+
+	private final TraceKeys traceKeys;
+	private final ObjectMapper objectMapper;
+
+	public HeaderBasedMessagingInjector(TraceKeys traceKeys, ObjectMapper objectMapper) {
+		this.traceKeys = traceKeys;
+		this.objectMapper = objectMapper;
+	}
+
+	@Override
+	public void inject(Span span, SpanTextMap carrier) {
+		Map<String, String> map = asMap(carrier);
+		if (span == null) {
+			if (!isSampled(map, TraceMessageHeaders.SAMPLED_NAME)) {
+				carrier.put(TraceMessageHeaders.SAMPLED_NAME, Span.SPAN_NOT_SAMPLED);
+				return;
+			}
+			return;
+		}
+		addHeaders(span, carrier);
+	}
+
+	private boolean isSampled(Map<String, String> initialMessage, String sampledHeaderName) {
+		return Span.SPAN_SAMPLED.equals(initialMessage.get(sampledHeaderName));
+	}
+
+	private void addHeaders(Span span, SpanTextMap textMap) {
+		addHeader(textMap, TraceMessageHeaders.TRACE_ID_NAME, Span.idToHex(span.getTraceId()));
+		addHeader(textMap, TraceMessageHeaders.SPAN_ID_NAME, Span.idToHex(span.getSpanId()));
+		if (span.isExportable()) {
+			addAnnotations(this.traceKeys, textMap, span);
+			Long parentId = getFirst(span.getParents());
+			if (parentId != null) {
+				addHeader(textMap, TraceMessageHeaders.PARENT_ID_NAME, Span.idToHex(parentId));
+			}
+			addHeader(textMap, TraceMessageHeaders.SPAN_NAME_NAME, span.getName());
+			addHeader(textMap, TraceMessageHeaders.PROCESS_ID_NAME, span.getProcessId());
+			addHeader(textMap, TraceMessageHeaders.SAMPLED_NAME, Span.SPAN_SAMPLED);
+		}
+		else {
+			addHeader(textMap, TraceMessageHeaders.SAMPLED_NAME, Span.SPAN_NOT_SAMPLED);
+		}
+		addHeader(textMap, TraceMessageHeaders.SPAN_HEADER, storeSpanAsJson(span));
```

Comment:
eh? we aren't literally sending the span in headers, right? that's basically the opposite of the intent of the in-band/reporting duo of dapper systems